catalog/spatial/csw/spatial-csw-endpoint/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswEndpoint.java

```diff
@@ -536,46 +554,103 @@ public TransactionResponseType transaction(CswTransactionRequest request) throws
     response.setVersion(CswConstants.VERSION_2_0_2);
 
     int numInserted = 0;
+    final Subject subject = SecurityUtils.getSubject();
     for (InsertAction insertAction : request.getInsertActions()) {
-      insertAction = transformInsertAction(insertAction);
-      CreateRequest createRequest = new CreateRequestImpl(insertAction.getRecords());
-      try {
-        CreateResponse createResponse = framework.create(createRequest);
-        if (request.isVerbose()) {
-          response.getInsertResult().add(getInsertResultFromResponse(createResponse));
-        }
+      final InsertAction transformInsertAction = transformInsertAction(insertAction);
+
+      List<Metacard> metacards = transformInsertAction.getRecords();
+      CompletionService<CreateResponse> completionService =
+          new ExecutorCompletionService<>(queryExecutor);
+
+      for (Metacard record : metacards) {
+        CreateRequest createRequest = new CreateRequestImpl(record);
+        Callable<CreateResponse> callable =
+            () -> {
+              try {
+                return framework.create(createRequest);
+              } catch (IngestException | SourceUnavailableException e) {
+                LOGGER.debug("Unable to insert record(s)", e);
+                throw new CswException(
+                    "Unable to insert record(s).",
+                    CswConstants.TRANSACTION_FAILED,
+                    transformInsertAction.getHandle());
+              }
+            };
+        Callable<CreateResponse> createCallable = subject.associateWith(callable);
+        completionService.submit(createCallable);
+      }
 
-        numInserted += createResponse.getCreatedMetacards().size();
-      } catch (IngestException | SourceUnavailableException e) {
-        LOGGER.debug("Unable to insert record(s)", e);
-        throw new CswException(
-            "Unable to insert record(s).",
-            CswConstants.TRANSACTION_FAILED,
-            insertAction.getHandle());
+      for (int i = 0; i < metacards.size(); i++) {
+        try {
+          Future<CreateResponse> completedFuture = completionService.take();
+
+          try {
+            CreateResponse futureResponse = completedFuture.get();
```

Comment:
Do we need to ensure that these return via a timeout or some other mechanism? Or can we guarantee this will return? (do we need to protect here or do we have guarantees that catalogframework will return?... i know we've had to fix bugs of infinite waits in catframework before, so i'd be weary to trust it 😛 100%)

Reply:
The previous call to the framework didn't have any timeout protections, so I'd like to leave this code as-is and rely on HTTP timeouts.